### DISCUSSION TITLE
Any way to update job data on fail?

### DISCUSSION THREAD

**Discussion by devalnor** [timgit/pg-boss#239] [Q&A]
I would like to know if there is a way to keep the job data updated (like a state).
I miss understood ```job.done(error, data)``` and I was thinking that data will update the job's data for a future cycle.
It's help me to keep a job state on the next retry.

**Comment by timgit** [comment]
Job data is immutable. However, a completion job is created with the data you pass as the arg to done(), or whatever you return from an async handler. You can use onComplete() or fetchComplete() to retrieve it.

**Reply by devalnor** [reply]
I suggest clarifying this information in the documentation. 
If job data is immutable, I have to roll back all my project...
Do you know another library that can keep a state object into a job?

**Reply by timgit** [reply]
I have a branch open that I intend to merge in the next week or 2 that adds an `output` prop onto the original job record. Once this is added, you will have access to this on your job if you add `includeMetadata` to your `subscribe()` or `fetch()` options.  Until then, you can add `onComplete: true` to your publish options and then get access to the output data with an `onComplete()` subscription or `fetchCompleted()` call.  Once you have the data, you can attach additional state (such as a retry count) and republish.  It's an extra subscription to set up, but it would resolve your goal in my opinion using the current version.
